File: KClosestElementsBinarySearch.java
// TC: O(log(n) + k)
// SC: O(1)
class Solution {
    public List<Integer> findClosestElements(int[] arr, int k, int x) {
        List<Integer> result = new ArrayList<>();
        if(arr == null || arr.length == 0) {
            return result;
        }

        if(arr.length == k) {
            for(int i = 0; i < k; i++) {
                result.add(arr[i]);
            }
            return result;
        }

        int l, r;
        l = 0;
        r = arr.length;
        while(l < r) {
            int m = l + (r - l)/ 2;
            if(arr[m] >= x) {
                r = m;
            } else {
                l = m + 1;
            }
        }

        l--;
        r = l + 1;
        while(r - l - 1 < k) {
            if(l == -1) {
                r++;
                continue;
            }
            if(r == arr.length || Math.abs(arr[l] - x) <= Math.abs(arr[r] - x)) {
                l--;
            } else {
                r++;
            }
        }

        for(int i = l + 1; i < r; i++) {
            result.add(arr[i]);
        }

        return result;
    }
}

File: KclosestElementsBinarySearch.py
'''Time Complexity: O(n-k)
Space Complexity: O(1)'''

class Solution(object):
            
    def findClosestElements(self, arr, k, x):
        """
        :type arr: List[int]
        :type k: int
        :type x: int
        :rtype: List[int]
        """
        #Optimized binary search method
        if len(arr)==0:
            return []
        #Binary Search
        n=len(arr)
        low=0
        high=n-k #start index will never go beyond n-k
        #Till the gap is greated than k, keep on reducing the gap
        while low < high :
            mid=low+(high-low)//2
            #assuming the start as mid
            start_dis=x-arr[mid]
            #therefore range will be from mid to mid+k
            end_dis=arr[mid+k]-x
            if start_dis > end_dis:
                low=mid+1
            else:
                high=mid
        
        #Now low will be the start index of the range
        result=[]
        for i in range(low, low+k):
            result.append(arr[i])
        return result
        
